app/components/SessionTimeoutHandler.tsx

```diff
@@ -0,0 +1,92 @@
+import {CiipPageComponent} from 'next-env';
+import {useRouter} from 'next/router';
+import React, {useEffect, useState} from 'react';
+import LogoutWarningModal from './LogoutWarningModal';
+
+interface Props {
+  pageComponent: CiipPageComponent;
+  modalDisplaySecondsBeforeLogout: number;
+}
+
+const SessionTimeoutHandler: React.FunctionComponent<Props> = ({
+  pageComponent,
+  modalDisplaySecondsBeforeLogout
+}) => {
+  const router = useRouter();
+
+  const [showModal, setShowModal] = useState(false);
+
+  // UNIX epoch (ms)
+  const [sessionExpiresOn, setSessionExpiresOn] = useState(Infinity);
+
+  const logoutOnSessionIdled = async () => {
+    router.push({
+      pathname: '/login-redirect',
+      query: {
+        redirectTo: router.asPath,
+        sessionIdled: true
+      }
+    });
+  };
+
+  const extendSession = async () => {
+    const response = await fetch('/extend-session');
+    if (response.ok) {
+      const timeout = Number(await response.json());
+      if (timeout > modalDisplaySecondsBeforeLogout) {
+        setShowModal(false);
+      }
+      setSessionExpiresOn(timeout * 1000 + Date.now());
+    }
+  };
+
+  useEffect(() => {
+    let timeoutId;
+
+    const checkSessionIdle = async () => {
+      if (!pageComponent.isAccessProtected) return;
+
+      const response = await fetch('/session-idle-remaining-time');
+      if (response.ok) {
```

Comment:
should we log them out if the response is not ok?

Reply:
I think if we forcefully redirect them on a temporary either server roadbump or a short internet outage, they might lose data...the current implementation would cancel all timeouts though. Maybe we can just do a "check again in 5 minutes if there is connectivity again"?

Reply:
If there's connectivity issues the promise is going to throw an exception here.

`and it will only reject on network failure or if anything prevented the request from completing.`
https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API

Then it looks like it would perhaps reach the sentry boundary?

Reply:
consensus reached: we will redirect the user to the logout page, since at that point things aren't functional